File: transfers/fraud/typologies/classic.py
from common.channels import FRAUD_CLASSIC, FRAUD_CYCLE
from common.transactions import Transaction
from math_models.amount_model import (
    FRAUD as FRAUD_MODEL,
    FRAUD_CYCLE as CYCLE_MODEL,
)
from transfers.factory import TransactionDraft

from ..engine import IllicitContext
from ..rings import Plan
from ..schedule import sample_burst_window, sample_timestamp
from .common import append_bounded_txn


def generate(
    ctx: IllicitContext,
    ring_plan: Plan,
    budget: int,
) -> list[Transaction]:
    """Generates a classic fraud-ring pattern with victim->mule, mule->fraud, and cycle hops."""
    if budget <= 0:
        return []

    rng = ctx.execution.rng
    out: list[Transaction] = []

    base_date, burst_days = sample_burst_window(
        rng,
        start_date=ctx.window.start_date,
        total_days=ctx.window.days,
        tail_padding_days=7,
        min_burst_days=2,
        max_burst_days=6,
    )

    for victim_acct in ring_plan.victim_accounts:
        if len(out) >= budget:
            break

        if ring_plan.mule_accounts and rng.coin(0.75):
            mule_acct = rng.choice(ring_plan.mule_accounts)
            ts = sample_timestamp(
                rng,
                base_date=base_date,
                max_days_offset=burst_days,
                min_hour=8,
                max_hour=22,
            )
            if not append_bounded_txn(
                ctx,
                out,
                budget,
                TransactionDraft(
                    source=victim_acct,
                    destination=mule_acct,
                    amount=FRAUD_MODEL.sample(rng),
                    timestamp=ts,
                    is_fraud=1,
                    ring_id=ring_plan.ring_id,
                    channel=FRAUD_CLASSIC,
                ),
            ):
                break

    for mule_acct in ring_plan.mule_accounts:
        if len(out) >= budget:
            break
        if not ring_plan.fraud_accounts:
            continue

        forwards = rng.int(2, 6)
        span_days = min(3, burst_days)

        for _ in range(forwards):
            if len(out) >= budget:
                break

            fraud_acct = rng.choice(ring_plan.fraud_accounts)
            ts = sample_timestamp(
                rng,
                base_date=base_date,
                max_days_offset=span_days,
                min_hour=0,
                max_hour=24,
            )
            if not append_bounded_txn(
                ctx,
                out,
                budget,
                TransactionDraft(
                    source=mule_acct,
                    destination=fraud_acct,
                    amount=FRAUD_MODEL.sample(rng),
                    timestamp=ts,
                    is_fraud=1,
                    ring_id=ring_plan.ring_id,
                    channel=FRAUD_CLASSIC,
                ),
            ):
                break

    nodes = ring_plan.fraud_accounts + ring_plan.mule_accounts
    if len(nodes) >= 3 and len(out) < budget:
        cycle_size = min(len(nodes), rng.int(3, 7))
        cycle_nodes = rng.choice_k(nodes, cycle_size, replace=False)
        passes = rng.int(2, 5)

        for _ in range(passes):
            if len(out) >= budget:
                break

            for idx, src_acct in enumerate(cycle_nodes):
                if len(out) >= budget:
                    break

                dst_acct = cycle_nodes[(idx + 1) % len(cycle_nodes)]
                ts = sample_timestamp(
                    rng,
                    base_date=base_date,
                    max_days_offset=burst_days,
                    min_hour=0,
                    max_hour=24,
                )
                if not append_bounded_txn(
                    ctx,
                    out,
                    budget,
                    TransactionDraft(
                        source=src_acct,
                        destination=dst_acct,
                        amount=CYCLE_MODEL.sample(rng),
                        timestamp=ts,
                        is_fraud=1,
                        ring_id=ring_plan.ring_id,
                        channel=FRAUD_CYCLE,
                    ),
                ):
                    break

    return out

File: src/transfers/fraud/typologies/classic.cpp
#include "phantomledger/transfers/fraud/typologies/classic.hpp"

#include "phantomledger/math/amounts.hpp"
#include "phantomledger/taxonomies/channels/types.hpp"
#include "phantomledger/transactions/draft.hpp"
#include "phantomledger/transfers/fraud/schedule.hpp"
#include "phantomledger/transfers/fraud/typologies/common.hpp"

#include <algorithm>
#include <cstdint>

namespace PhantomLedger::transfers::fraud::typologies::classic {

namespace {

class DraftMaker {
public:
  DraftMaker(std::uint32_t ringId, channels::Tag channel,
             std::int32_t chainId) noexcept
      : ringId_(ringId), channel_(channel), chainId_(chainId) {}

  [[nodiscard]] transactions::Draft make(const entity::Key &source,
                                         const entity::Key &destination,
                                         double amount,
                                         time::TimePoint ts) const {
    return transactions::Draft{
        .source = source,
        .destination = destination,
        .amount = amount,
        .timestamp = time::toEpochSeconds(ts),
        .isFraud = 1,
        .ringId = static_cast<std::int32_t>(ringId_),
        .channel = channel_,
        .chainId = chainId_,
    };
  }

private:
  std::uint32_t ringId_;
  channels::Tag channel_;
  std::int32_t chainId_;
};

void runVictimToMule(IllicitContext &ctx,
                     std::vector<transactions::Transaction> &out,
                     std::int32_t budget, const Plan &plan,
                     const BurstWindow &burst, const DraftMaker &maker) {
  random::Rng &rng = *ctx.execution.rng;
  for (const auto &victim : plan.victimAccounts) {
    if (static_cast<std::int32_t>(out.size()) >= budget) {
      return;
    }
    if (plan.muleAccounts.empty() || !rng.coin(0.75)) {
      continue;
    }
    const auto &mule = typologies::pickOne(rng, plan.muleAccounts);
    const auto ts = sampleTimestamp(rng, burst.baseDate, burst.durationDays,
                                    HourRange{.min = 8, .max = 22});
    if (!typologies::appendBoundedTxn(
            ctx, out, budget,
            maker.make(victim, mule, math::amounts::kFraud.sample(rng), ts))) {
      return;
    }
  }
}

void runMuleToFraud(IllicitContext &ctx,
                    std::vector<transactions::Transaction> &out,
                    std::int32_t budget, const Plan &plan,
                    const BurstWindow &burst, const DraftMaker &maker) {
  random::Rng &rng = *ctx.execution.rng;
  for (const auto &mule : plan.muleAccounts) {
    if (static_cast<std::int32_t>(out.size()) >= budget) {
      return;
    }
    if (plan.fraudAccounts.empty()) {
      continue;
    }
    const auto forwards = static_cast<std::int32_t>(rng.uniformInt(2, 7));
    const auto spanDays = std::min<std::int32_t>(3, burst.durationDays);
    for (std::int32_t i = 0; i < forwards; ++i) {
      if (static_cast<std::int32_t>(out.size()) >= budget) {
        return;
      }
      const auto &fraudAcct = typologies::pickOne(rng, plan.fraudAccounts);
      const auto ts = sampleTimestamp(rng, burst.baseDate, spanDays,
                                      HourRange{.min = 0, .max = 24});
      if (!typologies::appendBoundedTxn(
              ctx, out, budget,
              maker.make(mule, fraudAcct, math::amounts::kFraud.sample(rng),
                         ts))) {
        return;
      }
    }
  }
}

} // namespace

std::vector<transactions::Transaction>
generate(IllicitContext &ctx, const Plan &plan, std::int32_t budget) {
  std::vector<transactions::Transaction> out;
  if (budget <= 0) {
    return out;
  }

  random::Rng &rng = *ctx.execution.rng;

  const auto burst = sampleBurstWindow(rng, ctx.window.start, ctx.window.days,
                                       BurstShape{
                                           .tailPaddingDays = 7,
                                           .minDays = 2,
                                           .maxDays = 6,
                                       });

  const auto chainId = ctx.allocateChainId();
  const DraftMaker maker{plan.ringId, channels::tag(channels::Fraud::classic),
                         chainId};

  runVictimToMule(ctx, out, budget, plan, burst, maker);
  runMuleToFraud(ctx, out, budget, plan, burst, maker);

  return out;
}

} // namespace PhantomLedger::transfers::fraud::typologies::classic
